test/e2e/auth/audit.go

```diff
@@ -23,83 +23,623 @@ import (
 	"strings"
 
 	apiv1 "k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/test/integration/testserver"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apiserver/pkg/apis/audit/v1beta1"
 	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
 
+	"github.com/evanphx/json-patch"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 )
 
+var watchTestTimeout = int64(1)
+var auditTestUser = "kubecfg"
+
+type methodCall func()
+
 var _ = SIGDescribe("Advanced Audit [Feature:Audit]", func() {
 	f := framework.NewDefaultFramework("audit")
 
 	It("should audit API calls", func() {
 		namespace := f.Namespace.Name
 
-		// Create & Delete pod
-		pod := &apiv1.Pod{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "audit-pod",
-			},
-			Spec: apiv1.PodSpec{
-				Containers: []apiv1.Container{{
-					Name:  "pause",
-					Image: framework.GetPauseImageName(f.ClientSet),
-				}},
+		config, err := framework.LoadConfig()
+		framework.ExpectNoError(err, "failed to load config")
+		apiExtensionClient, err := clientset.NewForConfig(config)
+		framework.ExpectNoError(err, "failed to initialize apiExtensionClient")
+
+		crd := testserver.NewRandomNameCustomResourceDefinition(apiextensionsv1beta1.ClusterScoped)
+		crdName := strings.SplitN(crd.Name, ".", 2)[0]
+		crdNamespace := strings.SplitN(crd.Name, ".", 2)[1]
+
+		watchOptions := metav1.ListOptions{TimeoutSeconds: &watchTestTimeout}
+		patch, _ := json.Marshal(jsonpatch.Patch{})
+
+		testCases := []struct {
+			name   string
+			action func()
+			events []auditEvent
+		}{
+			{
+				"Pod",
+				func() {
+					pod := &apiv1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "audit-pod",
+						},
+						Spec: apiv1.PodSpec{
+							Containers: []apiv1.Container{{
+								Name:  "pause",
+								Image: framework.GetPauseImageName(f.ClientSet),
+							}},
+						},
+					}
+					updatePod := func(pod *apiv1.Pod) {}
+
+					f.PodClient().CreateSync(pod)
+
+					_, err := f.PodClient().Get(pod.Name, metav1.GetOptions{})
+					framework.ExpectNoError(err, "failed to get audit-pod")
+
+					podChan, err := f.PodClient().Watch(watchOptions)
+					for range podChan.ResultChan() {
+					}
+
+					framework.ExpectNoError(err, "failed to create watch for pods")
+
+					f.PodClient().Update(pod.Name, updatePod)
+					_, err = f.PodClient().List(metav1.ListOptions{})
```

Comment:
Wrong indents? Empty line seems more reasonable above this line, not below

Reply:
Fixed.